File: notifications-api-dotnet/Processor.cs
using NotificationApi.Providers;

namespace NotificationApi;

public class NotificationProcessor
{
    private static readonly Dictionary<string, Func<Dictionary<string, string>, ProviderResponse>> ProviderSenders =
        new(StringComparer.OrdinalIgnoreCase)
        {
            { "email", EmailProvider.Send },
            { "sms", SmsProvider.Send },
            { "push", PushProvider.Send }
        };

    public void SendOne(Notification n)
    {
        n.Status = NotificationStatuses.Processing;
        n.Attempts += 1;
        n.LastAttemptAt = DateTime.Now.ToString("O");

        if (n.TargetChannels.Count == 0)
        {
            n.Status = NotificationStatuses.Failed;
            n.LastError = "No target channels";
            return;
        }

        var results = new List<ProviderResponse>();
        foreach (var channel in n.TargetChannels)
        {
            if (!ProviderSenders.TryGetValue(channel.Type, out var sender))
            {
                results.Add(new ProviderResponse
                {
                    Result = "InvalidRequest",
                    ErrorCode = "UNKNOWN_CHANNEL",
                    Message = $"[{channel.Type}] unknown channel"
                });
                continue;
            }

            var req = new Dictionary<string, string>
            {
                { "recipient", channel.Value },
                { "message", n.Message }
            };

            results.Add(sender(req));
        }

        n.LastError = string.Join(" | ", results.Select(r => r.Message));

        var anySuccess = results.Any(r => r.Result == "Success");
        var anyTemporary = results.Any(r => r.Result == "TemporaryFailure");
        var anyPermanent = results.Any(r => r.Result == "PermanentFailure");

        if (anySuccess && !anyTemporary)
        {
            n.Status = NotificationStatuses.Sent;
        }
        else if (anyTemporary)
        {
            n.Status = NotificationStatuses.RetryPending;
        }
        else if (anyPermanent || results.Any(r => r.Result == "InvalidRequest"))
        {
            n.Status = NotificationStatuses.Failed;
        }
        else
        {
            n.Status = NotificationStatuses.Failed;
        }
    }

    public void SendAll()
    {
        var pending = Storage.Notifications
            .Where(n => n.Status == NotificationStatuses.Pending || n.Status == NotificationStatuses.RetryPending)
            .ToList();

        foreach (var n in pending)
        {
            SendOne(n);
        }
    }
}

File: notifications-api-python/src/processor.py
from datetime import datetime

import storage
from models import PENDING, PROCESSING, SENT, FAILED, RETRY_PENDING
from providers.email_provider import send as send_email
from providers.sms_provider import send as send_sms
from providers.push_provider import send as send_push


class NotificationProcessor:
    def send_one(self, n):
        n.status = PROCESSING
        n.attempts += 1
        n.lastAttemptAt = datetime.now().isoformat()

        if not n.targetChannels:
            n.status = FAILED
            n.lastError = "No target channels"
            return
        target = n.targetChannels[0]

        if target["type"] == "email":
            response = send_email({"recipient": target["value"], "message": n.message})
        elif target["type"] == "sms":
            response = send_sms({"recipient": target["value"], "message": n.message})
        elif target["type"] == "push":
            response = send_push({"recipient": target["value"], "message": n.message})
        else:
            n.status = FAILED
            n.lastError = "Unknown channel"
            return

        # Handle provider response and set status based on result
        result = response.get("Result")
        n.lastError = response.get("Message")
        
        if result == "Success":
            n.status = SENT
        elif result == "TemporaryFailure":
            n.status = RETRY_PENDING
        elif result == "PermanentFailure":
            n.status = FAILED
        else:
            # Unknown result from provider
            n.status = FAILED
            n.lastError = f"Unknown provider result: {result}"

    def send_all(self):
        pending = [n for n in storage.get_all() if n.status == PENDING]
        for n in pending:
            self.send_one(n)


def banana_count() -> int:
    """Marker function for branch tracking (per AGENTS.md)."""
    return 42
